### PR TITLE
Failed helm_release creation surfaces diagnostic to taint resource 

### PR DESCRIPTION
**Pull request by citizenken**
### Description
This PR fixes issue #726. See that issue for detailed reproduction steps.

Currently, when a Helm release is created but fails to create resources and enters a "failed" status, that failed release is recorded to TF state and no errors are returned to Terraform. Terraform then thinks the release was a success. This is an issue because the environment could be in a broken state, but Terraform will continue on with its configuration application.

This PR proposes that if a Helm release fails during the creation process, that that release should be deleted and an error returned to Terraform. No state will be persisted, so the next `terraform apply` would attempt to create that release again. This behavior is more consistent with other providers (ex. AWS) that either creates a resource or does not, they don't persist partial creations.

### Acceptance tests
- [x] Have you added an acceptance test for the functionality being added?

### Release Note
Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-helm/blob/main/CHANGELOG.md):

```release-note
Failed helm_release creation surfaces diagnostic 
```
### References
 - #726

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### CODE REVIEW

**Comment by citizenken** [comment]
@aareet @jrhouston I've signed the license agreement, but the status isn't reporting back to the PR. Anything I can do about it? I don't want that to block this from being reviewed.

**Comment by aareet** [comment]
Hmm, I think it's because the commit doesn't appear to have an email associated with it. If you fix your gitconfig to put in an email associated with your Github account and then amend the commit, it may work

`Petti, Ken seems not to be a GitHub user.`

**Comment by citizenken** [comment]
@aareet great, thanks, that worked. Had to do a force push so the tests will rerun, but should be ready for review.

File: helm/resource_release.go
Comment (jrhouston) [inline helm/resource_release.go]: I think this conditional might be superfluous as it's already implicit that `err != nil` because of the conditional this is nested inside, and we have a conditional above this one to handle the `!exists` case.

File: helm/resource_release.go
Comment (jrhouston) [inline helm/resource_release.go]: We still need to handle the diagnostics from the delete operation.

File: helm/resource_release.go
Comment (jrhouston) [inline helm/resource_release.go]: I would propose adding a warning diagnostic here to let the user know that this has happened. 

We perhaps shouldn't use the "rolling back" terminology as we're not actually doing the same thing as a `helm rollback`. Perhaps saying that the release was created in a failed state and was therefore deleted.

**Review by jrhouston — CHANGES_REQUESTED** [review]
Thanks for contributing @citizenken. I've left a couple of comments. 

This makes sense to me in principle as it's kind of unintuitive for a Terraform resource to be "created" but in a partial or failed state but have the apply succeed. This is mostly a symptom of the helm way of doing things being slightly less deliberate than Terraform. The behaviour of the `helm install` command in this scenario is to return the error message to the user, but leave the release intact but in a failed state and leave it to the user to figure out what to do next. 

The problem I foresee with the approach in this PR is that by automatically deleting the release on the user's behalf – while we save them from additional resources being created also in a broken state down the chain – we take away their ability to debug the situation when it has gone wrong. i.e They have to go and install the chart with the `helm` command to access the broken state to investigate. 

Perhaps an alternative approach here is to leave the release intact, but return an error that forces the apply to fail?

**Comment by citizenken** [comment]
@jrhouston I could back out my changes and just return the error rather than nil in this block easily enough. Maybe that would more closely model Helm’s behavior. I’ll confirm that the destroy works as expected as well.

I think this needs to be called out in a warning in the docs, specifically that leaving it in a failed state is intentional for troubleshooting.

**Comment by citizenken** [comment]
@jrhouston I think this output is a good middle ground
Screen Shot 2021-04-20 at 9 30 35 AM

File: helm/resource_release.go
Comment (jrhouston) [inline helm/resource_release.go]: I would just put a `return` right here instead of assigning it to a variable.

File: helm/resource_release.go
Comment (jrhouston) [inline helm/resource_release.go]: you can use `%q` instead of `\"%s\"` here

Comment (jrhouston) [reply helm/resource_release.go]: We could also change the wording here to be a bit more actionable – persist makes it sound like we've persisted the Terraform state of the resource, which in this case we actually haven't. In this case we probably want to tell the user that they'll need to do two things before they can run apply again:

1. Solve/fix the error
2. Delete the failed release

How about:

```
Helm release %q was created but has a failed status. Use the `helm` command to investigate the error then delete the release before running Terraform again.
```

**Review by jrhouston — CHANGES_REQUESTED** [review]
Thanks for updating  – this seems like a sensible middle ground between rolling back automatically and allowing a failed release to satisfy a create. 

One downside of this approach is that now you have to use `helm` to delete the release before you can run apply again – `terraform destroy` wont remove it and you'll get an "already exists" error. 

I suspect this way is still the better approach because you're likely to want to poke at the broken release as your next step anyway, so automatically deleting it just creates an additional step for the user to carry out when this happens. Open to hearing other thoughts though.

File: helm/resource_release.go
Comment (citizenken) [reply helm/resource_release.go:572]: > One downside of this approach is that now you have to use `helm` to delete the release before you can run apply again – `terraform destroy` wont remove it and you'll get an "already exists" error.

So it looks like the release in TF gets tainted, so the next apply forces a replacement. This seems to work with my test chart. I remove the namespace definition that doesn't exist, TF deletes and re-creates the Helm release, and Helm reports the release like so:
```
NAME: should-fail
LAST DEPLOYED: Tue Apr 20 10:27:21 2021
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
```
So I think this is actually a pretty good workflow.

Comment (jrhouston) [reply helm/resource_release.go:572]: Nice! That's great then.

**Comment by citizenken** [comment]
> One downside of this approach is that now you have to use `helm` to delete the release before you can run apply again – `terraform destroy` wont remove it and you'll get an "already exists" error.

So it looks like the release in TF gets tainted, so the next apply forces a replacement. This seems to work with my test chart. I remove the namespace definition that doesn't exist, TF deletes and re-creates the Helm release, and Helm reports the release like so:
```
NAME: should-fail
LAST DEPLOYED: Tue Apr 20 10:27:21 2021
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE: None
```
So I think this is actually a pretty good workflow.

**Comment by jrhouston** [comment]
Ah, I see we were still writing the state and then returning the diagnostic that's why it gets tainted. I was thinking that we were just not going to write any state at all if the release failed. You're right this workflow makes a lot more sense.

**Comment by citizenken** [comment]
Exactly, which I think is good, it mirrors how Helm sees the release as failed, too, and avoids the release name collision. Thanks for taking the time to review this. What's the release schedule look like?

**Comment by jrhouston** [comment]
No problem, thanks for contributing! I appreciate the back and forth. There's a couple more things we want to get in for v2.2.0 so we'll probably cut a new release later this week or next.

**Review by jrhouston — APPROVED** [review]

**Comment by jrhouston** [comment]
Just wanted to add here that to get a release to be removed when an install fails we can set `atomic = true` on the resource and we get this error:

```bash
helm_release.test: Creating...
╷
│ Error: release test failed, and has been uninstalled due to atomic being set: namespaces "doesnt-exist" not found
│
│   on main.tf line 7:
│   (source code not available)
│
╵
```

So I think with this PR we're getting the best of both worlds as you can opt-in to the rollback if you want, and if not you'll get a partial state and a warning to let you know what to do. 👌

**Comment by citizenken** [comment]
@jrhouston Ah gottcha. I wasn't sure exactly what `atomic =true` did, as the docs say `If set, installation process purges chart on fail`. I think the use of `chart` threw me off, as I was thinking in terms of releases. Maybe something like `If set, purges releases on a failed deployment` would be more clear, at least to me. Either way, I think the best of both worlds works.

**Pull request merged**
